---
title: "ILoReg package manual"
author: "Johannes Smolander"
date: "`r Sys.Date()`"
output: 
  BiocStyle::html_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{ILoReg package manual}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding[utf8]{inputenc}
---

<style>
body {
text-align: justify}
</style>


```{r knitr-options, echo=FALSE, message=FALSE, warning=FALSE}
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5, dev = 'png')
```


# Introduction to the method
`ILoReg` (abbr. for **I**terative **Lo**gistic **Reg**ression) is a tool for high-precision cell type identification from single-cell RNA-seq (scRNA-seq) data. High-precision refers to the ability of `ILoReg` to identify subpopulations with subtle transcriptomic differences. In our study ([Article title](https://gitlab.utu.fi/pajosm/iloreg)), we showed that `ILoReg` was able to identify, by both unsupervised clustering and visually, immune cell types that other scRNA-seq data analysis pipelines were not able to find. Moreover, `ILoReg` can identify subpopulations that are differentiable by only a single gene.

The figure below depicts the workflows of `ILoReg` and a feature selection -based pipeline, which is commonly used by many cell type identification methods, e.g. Seurat.

![*Figure: Analysis workflows of `ILoReg` and a feature-selection based approach*](figure.png){width=100%}

Unlike most scRNA-seq data analysis pipelines, `ILoReg` does not reduce the dimensionality of the gene expression matrix by feature selection. Instead, it performs probabilistic feature extraction, in which the Iterative Clustering Projection (ICP) clustering algorithm is run *L* times, which yields *L* *k*-dimensional probability matrices that contain the new features. ICP is a novel clustering algorithm that iteratively seeks a clustering of size *k* that maximizes the adjusted Rand index (ARI) between the clustering $C$ and its projection $C'$ by L1-regularized logistic regression: $$\displaystyle \arg \max_C ARI(C,C')$$ The *L* probability matrices are then merged and transformed by the principal component analysis (PCA) into a lower dimension *p*. The second and final clustering step is performed using hierarhical clustering by the Ward's method, from which the user can efficiently (~1 s with 3,000 cells) select a clustering of size *K*. Two-dimensional visualization is supported using two popular nonlinear dimensionality reduction methods: *t*-Distributed Stochastic Neighbor Embedding (t-SNE) and Uniform Manifold Approximation and Projection (UMAP).


# Installation

The latest version of `ILoReg` can be downloaded from GitHub using utilities from devtools R package.


```{r eval = FALSE}
library(devtools)

creds = git2r::cred_ssh_key("~/../.ssh/id_rsa.pub",
                            "~/../.ssh/id_rsa")
devtools::install_git("gitlab@gitlab.utu.fi:pajosm/iloreg.git",
                      credentials = creds)
```


# Example: Peripheral Blood Mononuclear Cells

In the following, we go through the different steps of `ILoReg`'s workflow and demonstrate it using a Peripheral Blood Mononuclear Cell (PBMC) dataset. The dataset (pbmc3k) can be downloaded from the following link: [GitHub](https://support.10xgenomics.com/single-cell-gene-expression/datasets)



### Setup a SingleCellExperiment object and prepare it for ILoReg analysis

The only required input for `ILoReg` is a normalized gene expression matrix with genes/features in rows and cells/samples in columns. The input can be of class *matrix*, *data.frame* or *dgCMatrix* and the matrix is then transformed into a sparse object of class *dgCMatrix*. **Please note that the method has been designed to work with sparse data, i.e. with lots of zero values. If for example the features of your dataset have been standardized, the method is likely to work much slower and require more RAM.**



```{r}
# data(pbmc3k_1000)

library(SingleCellExperiment)
sce <- SingleCellExperiment(assays = list(logcounts = pbmc3k_1000))
```


```{r}
library(ILoReg)
# sce <- PrepareILoReg(sce)
```


### Run ICP clustering algorithm *L* times

Next comes by far the most computationally demanding part of the pipeline, 
in which *L* ICP solutions are computed with different random seeds. This step lasted ~2 hr with three threads on a laptop. 

In the following, we give a brief summary of the parameters.

* $k$: The number of clusters in ICP. Default $15$. A lower value generates less cluster probabilities. Increasing $k$ is the easiest way to increase the complexity of the outcome, i.e. more subpopulations with subtle differences emerge.
* $C$: A positive real number, the cost of constraints violation in the L1-regularized logistic regression model from the [LIBLINEAR library](https://www.csie.ntu.edu.tw/~cjlin/liblinear/): $$\displaystyle \min_w {\Vert w \Vert}_1 + C \sum_{i=1}^{l} \log (1+ e^{-y_i w^T w})$$ Default 0.3. Decreasing leads to more stringent feature selection, i.e. less genes are selected that are used to train the projection classifier.
* $d$: A real number greater than $0$ and smaller than $1$ that determines how many cells $n$ are down- or oversampled from each cluster into the training data. ($d$ in $n=Nd/k$), where $N$ is the total number of cells. Default $0.3$. Our recommendation is to not increase $d$ above $0.5$, as this will lead to more dissimilar ICP runs, which in turn induces less variation into the consensus solution. 
* $r$: A positive integer that denotes the maximum number of reiterations performed until the ICP algorithm stops. Default $5$. Increasing recommended with a significantly larger sample size (tens of thousands).
* $L$: Number of ICP runs. Default $200$. 
* $type$: "L1" or "L2". "L2" denotes L2-regularization (Tikhonov/Ridge regularization), which was not investigated in the manuscript. Default "L1".
* $seed$: To make the results reproducible, a positive integer that specifies the random seed to be set before randomly generating random seeds for each parallel task. Default $1$.
* $threads$: The number of threads to use in parallel computing. Default $0$, which uses all available threads but one. Parallelization can be disabled with $1$. 


```{r eval = FALSE}
# Run ICP L times. This is by far the slowest step of the workflow and parallel processing can be used to greatly speed it up.
tictoc::tic()
set.seed(1)
sce <- RunParallelICP(object = sce, k = 15, d = 0.3, L = 200, r = 5, C = 0.3, 
                      reg.type = "L1",
                      threads = 0)
tictoc::toc()
# At this point saving the results is recommended.
saveRDS(sce,file = "~/pbmc3k_1000_iloreg.rds")
```

```{r}
sce <- readRDS(file = "~/pbmc3k_1000_iloreg.rds")
```



### Quality control

Next, we consider two quality control measures of ICP that can aid detecting issues with the ICP analysis.

First, the terminal projection accuracy that denotes the adjusted Rand index (ARI) between the clustering and its logistic regression projection at the final epoch (iteration) of ICP. Projection accuracy 1 would mean the classifier was able to predict the cluster identity of all cells correctly. Higher accuracy can be achieved using lower values of $k$, $C$ and higher values of $d$ and $r$. 

Second, the average pairwise ARI that measures how similar the different ICP runs are on average respect to each other. Higher $d$ leads to lower average pairwise ARI and smaller $k$ to higher average pairwise ARI. The relation of $C$ and $r$ with the average pairwise ARI is not as unambiguous as with the projection accuracy.

```{r, fig.height=4, fig.width=9, out.width = '100%',fig.align = "center"}
# Draw violin plots of the terminal clustering projection accuracy, the number of epochs and the average pairwise ARI.
# Each data point represents a ICP run.
VisualizeQC(sce,return.plot = F)
```


### PCA transformation of the joint probability matrix

The $L$ probability matrices are merged into a joint probability matrix, which is transformed into a lower dimension $p$ by principal component analysis (PCA). At this point, the user can filter out ICP runs that achieved lower terminal projection accuracy. Moreover, scaling the probabilities to unit-variance is supported.

```{r}
sce <- RunPCA(sce,p=50,scale = FALSE,threshold = 0)
```


PCA transformation requires the user to select the number of principal components, for which we selected default value $p=50$. The elbow plot is commonly used to seek a point of inflection, from which proximity the user selects $p$.

```{r, fig.height=5, fig.width=7, out.width = '100%',fig.align = "center"}
PCAElbowPlot(sce)
```


### Nonlinear dimensionality reduction

To visualize the data in two-dimensional space, nonlinear dimensionality reduction is performed using either t-SNE or UMAP. 
The input for this step is the $p$-dimensional PCA-rotated data matrix.

```{r}
sce <- RunUMAP(sce)
sce <- RunTSNE(sce,perplexity=30)
```


### Hierarchical clustering using Ward's agglomeration

The $p$-dimensional PCA-rotated data matrix is clustered hierarchically using the Ward's method.

```{r}
sce <- HierarchicalClustering(sce)
```


### Gene expression visualization

Visualize expression of CD3D (T cells), CD79A (B cells), CST3 (monocytes, dendritic cells, megakaryocytes), FCER1A (myeloid dendritic cells).
```{r, fig.height=10, fig.width=12, out.width = '100%'}
GeneScatterPlot(sce,c("CD3D","CD79A","CST3","FCER1A"),dim.reduction.type = "umap")
GeneScatterPlot(sce,c("CD3D","CD79A","CST3","FCER1A"),dim.reduction.type = "tsne")

```

### Selecting a clustering of size $K$

Manually select $K=40$ clusters from the dendrogram. This clustering is saved to *clustering.manual* slot and is overwritten every time this function is called.
```{r}
sce <- SelectKClusters(sce,K=20)
```


tSNE and UMAP transformations along with the optimal and manual clusterings.
```{r FigUMAP_TSNE, echo=TRUE, fig.height=5, fig.width=12, out.width = '100%'}
# Use plot_grid function from the cowplot R package to combine to plots into one.
library(cowplot)
plot_grid(ClusteringScatterPlot(sce,
                                dim.reduction.type = "umap",
                                return.plot = T,
                                title = "UMAP",
                                show.legend=FALSE),
          ClusteringScatterPlot(sce,
                                dim.reduction.type = "tsne",
                                return.plot = T
                                ,title="t-SNE",
                                show.legend=FALSE)
)

```


### Merging clusters

Identifying rare cell types, e.g. megakaryocytes or plasmacytoid dendritic cells, might require selecting a high value of $K$, but this in turn can form clusters that are not meaningful. Therefore, `ILoReg` provides a simple function for combining clusters.
```{r Fig0, echo=TRUE, fig.height=10, fig.width=12, out.width = '100%', eval = FALSE}
# Visualization before merging
tsne_before <- ClusteringScatterPlot(sce,dim.reduction.type = "tsne",return.plot = T,title = "Before merging clusters",show.legend = FALSE)
umap_before <- ClusteringScatterPlot(sce,dim.reduction.type = "umap",return.plot = T,title = "Before merging clusters",show.legend = FALSE)
# Merge multiple sets of clusters
sce <- MergeClusters(sce,clusters.to.merge = c(17,38),new.name = "m1")
sce <- MergeClusters(sce,clusters.to.merge = c(34,7,8),new.name = "m2")
sce <- MergeClusters(sce,clusters.to.merge = c(33,35),new.name = "m3")
sce <- MergeClusters(sce,clusters.to.merge = c(4,9),new.name = "m4")
sce <- MergeClusters(sce,clusters.to.merge = c(27,20,18,30,31,11),new.name = "m5")
sce <- MergeClusters(sce,clusters.to.merge = c(22,36,16,12),new.name = "m6")
# Visualization after merging
tsne_after <- ClusteringScatterPlot(sce,dim.reduction.type = "tsne",return.plot = T,title = "After merging clusters",show.legend = FALSE)
umap_after <- ClusteringScatterPlot(sce,dim.reduction.type = "umap",return.plot = T,title = "After merging clusters",show.legend = FALSE)
plot_grid(tsne_before,tsne_after,umap_before,umap_after)
```


### Identification of gene markers

`ILoReg` provides functions for identifying gene markers for clusters. This is done by differential expression analysis, where cells from each cluster are compared against the rest of the cells. The only statistical test supported at the moment is the the Wilcoxon rank-sum test (aka Mann-Whitney U test). The p-values are corrected for multiple comparisons using the Bonferroni method. 
```{r, eval = FALSE}

gene_markers <- FindAllGeneMarkers(sce,
                                   clustering.type = "manual",
                                   test = "wilcox",
                                   log2fc.threshold = 0.25,
                                   min.pct = 0.25,
                                   min.diff.pct = NULL,
                                   pseudocount.use = 1,
                                   min.cells.group = 3,
                                   return.thresh = 0.01,
                                   only.pos = FALSE,max.cells.per.cluster = NULL)

```


### Selecting top genes based on log2 fold-change

Select top 10 and top 1 genes based on log2 fold-change and Bonferroni adjusted p-value.
```{r, eval = FALSE}
top10_log2FC <- SelectTopGenes(gene_markers,top.N = 10,criterion.type = "log2FC",reverse = FALSE)
top1_log2FC <- SelectTopGenes(gene_markers,top.N = 1,criterion.type = "log2FC",reverse = FALSE)
top10_adj.p.value <- SelectTopGenes(gene_markers,top.N = 10,criterion.type = "adj.p.value",reverse = TRUE)
top1_adj.p.value <- SelectTopGenes(gene_markers,top.N = 1,criterion.type = "adj.p.value",reverse = TRUE)
```



### Gene expression scatter plot

Visualize gene expression levels of cells over t-SNE or UMAP visualization.

```{r FigTop1Scatter, echo=TRUE, fig.height=16, fig.width=20, out.width = '100%', eval = FALSE}
GeneScatterPlot(sce,genes = unique(top1$gene),dim.reduction.type = "tsne",point.size = 0.7)
```


```{r FigTop1Scatter2, echo=TRUE, fig.height=16, fig.width=20, out.width = '100%', eval = FALSE}
GeneScatterPlot(sce,genes = unique(top1$gene),dim.reduction.type = "umap",point.size = 0.7)
```


### Gene marker heatmap

GeneHeatmap function enables visualizing gene markers in a heatmap, where cells and genes are grouped by the clustering.

```{r FigHM, echo=TRUE, fig.height=15, fig.width=15, out.width = '100%', eval = FALSE}
GeneHeatmap(sce,clustering.type = "manual",gene.markers = top10_log2FC)

```


### Identification of differentially expressed genes between two arbitrary sets of clusters

```{r, eval = FALSE}
gene_markers_IGLL5_positive_b_cell_VS_negative <- FindGeneMarkers(sce,
                                                                  clusters.1 = c(29,15,14),
                                                                  clusters.2 = c(26,2),
                                                                  logfc.threshold = 0,
                                                                  min.pct = 0,
                                                                  return.thresh = 0.01,
                                                                  only.pos = TRUE)

gene_markers_TCL1A_positive_b_cell_VS_negative <- FindGeneMarkers(sce,
                                                                  clusters.1 = c(26,29,15),
                                                                  clusters.2 = c(14,2),
                                                                  logfc.threshold = 0,
                                                                  min.pct = 0,
                                                                  return.thresh = 0.01,
                                                                  only.pos = TRUE)

```



### Visualization of gene expression grouped by clusters with violin plots

Violin plots provide a nice way to visualize gene expression across clusters. 

```{r FigVlnPlot, echo=TRUE, fig.height=5, fig.width=12, out.width = '100%'}
VlnPlot(sce,clustering.type = "manual",genes = c("CD3D","CD79A","CST3"),return.plot = F)
```



### Saving results

Saving the results at the end of the workflow is recommended.

```{r}
saveRDS(sce,file = "~/pbmc3k_sce_final.rds")
```



# Additional functionality

`ILoReg` provides additional functionality for performing tasks, which are commonly required in scRNA-seq data analysis.

### Estimating optimal number of clusters

Estimation of the optimal number of clusters is done by calculating the average silhouette value across the cells for clusterings with different $K$ values in a range (e.g. $[2,50]$). The clustering acquired using $K$ with the highest average silhouette is saved to *clustering.optimal* slot.

```{r, fig.height=5, fig.width=7, out.width = '100%',fig.align = "center"}
sce <- CalculateSilhouetteInformation(sce,K.start = 2,K.end = 50)
SilhouetteCurve(sce,return.plot = F)
```


### Renaming clusters

Rename all clusters at once with RenameAllClusters function.
```{r FigRenaming, echo=TRUE, fig.height=10, fig.width=12, out.width = '100%', eval = FALSE}
sce <- SelectKClusters(sce,K=20)
tsne_before <- ClusteringScatterPlot(sce,dim.reduction.type = "tsne",title = "Before renaming",show.legend = FALSE,return.plot = TRUE)
umap_before <- ClusteringScatterPlot(sce,dim.reduction.type = "umap",title = "Before renaming",show.legend = FALSE,return.plot = TRUE)
sce <- RenameAllClusters(sce,new.cluster.names = LETTERS[1:sce@K.manual])
tsne_after <- ClusteringScatterPlot(sce,dim.reduction.type = "tsne",title = "After renaming",show.legend = FALSE,return.plot = TRUE)
umap_after <- ClusteringScatterPlot(sce,dim.reduction.type = "umap",title = "After renaming",show.legend = FALSE,return.plot = TRUE)
plot_grid(tsne_before,tsne_after,umap_before,umap_after)
```


Rename one cluster with RenameCluster function.
```{r FigRenaming2, echo=TRUE, fig.height=10, fig.width=12, out.width = '100%',eval = FALSE}
sce <- SelectKClusters(sce,K=20)
tsne_before <- ClusteringScatterPlot(sce,dim.reduction.type = "tsne",title = "Before renaming",show.legend = FALSE,return.plot = TRUE)
umap_before <- ClusteringScatterPlot(sce,dim.reduction.type = "umap",title = "Before renaming",show.legend = FALSE,return.plot = TRUE)
sce <- RenameCluster(sce,old.cluster.name = 1,new.cluster.name = "A")
tsne_after <- ClusteringScatterPlot(sce,dim.reduction.type = "tsne",title = "After renaming",show.legend = FALSE,return.plot = TRUE)
umap_after <- ClusteringScatterPlot(sce,dim.reduction.type = "umap",title = "After renaming",show.legend = FALSE,return.plot = TRUE)
plot_grid(tsne_before,tsne_after,umap_before,umap_after)
```


### Visualize with custom cluster labels

```{r FigAnnotation, echo=TRUE, fig.height=5, fig.width=6, out.width = '100%',fig.align = "center",eval = FALSE}
# Select a clustering with K=20 clusters
sce <- SelectKClusters(sce,K=20)
# Generate a random annotation, in which cell has 90% change of having the same cluster as in ILoReg's clustering
custom_annotation <- unlist(lapply(1:ncol(sce@normalized.data),function(x) {prob <- rep(0.1/(sce@K.manual-1),sce@K.manual) ; prob[levels(sce@clustering.manual)==sce@clustering.manual[x]] <- 0.9 ; sample(levels(sce@clustering.manual),size = 1,prob = prob)}))
# Visualize the annotation
AnnotationScatterPlot(sce,annotation = custom_annotation,return.plot = F,dim.reduction.type = "tsne",show.legend=FALSE)
```



### Dropout curve

The dropout curve can be used to infer whether a gene is likely to be differentially expressed based on its abnormally high average expression level considering its dropout rate, i.e. fraction cells not expressing the gene. Genes that are differentially expressed in a population of cells are likely to be located above the expected curve [1].

We investigate how four B cell marker genes are situated in the dropout curve. *CD79A* is expressed in all B cells, but the other three genes appear to be differentially expressed.

```{r, echo=TRUE, fig.height=10, fig.width=12, out.width = '100%'}
GeneScatterPlot(sce,genes = c("CD79A","IGLL5","TCL1A","VPREB3"),dim.reduction.type = "tsne",point.size = 0.7)
```



From the dropout curve below we can indeed see, that *CD79A* is located close to of the center of the curve. *VPREB3* is slightly above the curve, whereas *TCL1A* and *IGLL5* are positioned clearly above the center of the curve.

```{r, echo=TRUE, fig.height=6, fig.width=8, out.width = '100%',fig.align = "center"}
# Select a clustering with K=20 clusters
sce <- SelectKClusters(sce,K=20)
# Investigate dropout rates and average non-zero expression of a few potential differentially genes in B cells
GeneDropoutRatePlot(sce,genes = c("IGLL5","CD79A","TCL1A","VPREB3"),return.plot = F,clusters = c(17,11,16,19))
```




# References 

1. Andrews, Tallulah S., and Martin Hemberg. "M3Drop: Dropout-based feature selection for scRNASeq." Bioinformatics (2018).

# Contact information

If you find bugs from `ILoReg` or have suggestions on how to improve our pipeline, please contact us in [GitHub](https://gitlab.utu.fi/pajosm/iloreg). 

